clouddriver-cf/src/main/groovy/com/netflix/spinnaker/clouddriver/cf/provider/agent/ClusterCachingAgent.groovy

```diff
@@ -0,0 +1,526 @@
+/*
+ * Copyright 2016 Pivotal, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.cf.provider.agent
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.frigga.Names
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.cats.agent.*
+import com.netflix.spinnaker.cats.cache.CacheData
+import com.netflix.spinnaker.cats.cache.DefaultCacheData
+import com.netflix.spinnaker.cats.provider.ProviderCache
+import com.netflix.spinnaker.clouddriver.cache.OnDemandAgent
+import com.netflix.spinnaker.clouddriver.cache.OnDemandMetricsSupport
+import com.netflix.spinnaker.clouddriver.cf.CloudFoundryCloudProvider
+import com.netflix.spinnaker.clouddriver.cf.cache.Keys
+import com.netflix.spinnaker.clouddriver.cf.config.CloudFoundryConstants
+import com.netflix.spinnaker.clouddriver.cf.model.CloudFoundryService
+import com.netflix.spinnaker.clouddriver.cf.provider.CloudFoundryProvider
+import com.netflix.spinnaker.clouddriver.cf.provider.ProviderUtils
+import com.netflix.spinnaker.clouddriver.cf.security.CloudFoundryAccountCredentials
+import com.netflix.spinnaker.clouddriver.cf.utils.CloudFoundryClientFactory
+import org.cloudfoundry.client.lib.CloudFoundryException
+import org.cloudfoundry.client.lib.CloudFoundryOperations
+import org.cloudfoundry.client.lib.domain.*
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE
+import static com.netflix.spinnaker.clouddriver.cf.cache.Keys.Namespace.*
+
+/**
+ * @author Greg Turnquist
+ */
+class ClusterCachingAgent implements CachingAgent, OnDemandAgent, AccountAware {
+
+  final static Logger log = LoggerFactory.getLogger(ClusterCachingAgent)
+
+  @Deprecated
+  private static final String LEGACY_ON_DEMAND_TYPE = 'CloudFoundry' // not relevant for CF
+
+  private static final String ON_DEMAND_TYPE = 'ServerGroup'
+
+  private static final TypeReference<Map<String, Object>> ATTRIBUTES = new TypeReference<Map<String, Object>>() {}
+
+  static final Set<AgentDataType> types = Collections.unmodifiableSet([
+      AUTHORITATIVE.forType(SERVER_GROUPS.ns),
+      AUTHORITATIVE.forType(CLUSTERS.ns),
+      AUTHORITATIVE.forType(APPLICATIONS.ns),
```

Comment:
An application can span multiple providers, so this should be `INFORMATIVE.forType(APPLICATIONS.ns)`